### PR TITLE
Make sure that all file loading happens via SourceMap

### PR DESCRIPTION
**Pull request by matklad**
That way, callers don't need to repeat "let's add this to sm manually
for tracking dependencies" trick.

It should make it easier to switch to using `FileLoader` for binary
files in the future as well

cc #62948 

r? @petrochenkov

### CODE REVIEW

File: src/libsyntax_ext/source_util.rs [hunk 1]
```diff
@@ -114,20 +112,17 @@ pub fn expand_include_str(cx: &mut ExtCtxt<'_>, sp: Span, tts: &[tokenstream::To
...
-    match fs::read_to_string(&file) {
-        Ok(src) => {
-            let interned_src = Symbol::intern(&src);
-
-            // Add this input file to the code map to make it available as
-            // dependency information
-            cx.source_map().new_source_file(file.into(), src);
-
-            base::MacEager::expr(cx.expr_str(sp, interned_src))
+    match cx.source_map().load_binary_file(&file) {
```
Comment (matklad) [inline src/libsyntax_ext/source_util.rs]: I am not sure whether `include_str` should normalize file contents... This approach does not do normalization, just like the previous version.

Comment (petrochenkov) [reply src/libsyntax_ext/source_util.rs]: I'm pretty sure it should.
The content included by `include_str` is still text / source code, if it's a long multi-line string encoded on Windows and stored in a file with UTF-8 BOM, then all those `\r\n`s and BOM shouldn't get into the string for the same reason why any other (non-`include_bytes`) inclusion method filters them away from string literals.

Am I right that https://github.com/rust-lang/rust/pull/62948 applied the normalization to `include_str` as well and crater found no regression?
This should give as the right to make the change (in this PR or in #62948, not much difference).

**Comment by petrochenkov** [comment]
r=me if you don't plan to address the `include_str` issue in this PR.

**Comment by matklad** [comment]
>Am I right that #62948 applied the normalization to include_str as well and crater found no regression?

I think that's not the case

* it was a check run, so it couldn't have detected this.
* similar to `include_bytes`, normalization was applied only to the data stored in the `SourceMap`, the actual included string was not normalized. 


I've added commit that adds normalization to `include_str`, which currently means BOM removal. It seems reasonable to me because programs (ie, compiler) are supposed to work the same regardless of the presence of BOM. Moreover, if we don't strip BOM, there's a high chance that it ends up being in the middle of some text stream produced by the program.

This is technically a breaking change, so we should probably FCP it?

**Comment by matklad** [comment]
>we know that code tested by crater doesn't include_str files with isolated \r though.

Actually, even this is not true:

https://github.com/nabijaczleweli/codepage-437/blob/2e17570377121dbd2bab5adebfcb46172804b95b/tests/cp437_control/mod.rs#L7

**Comment by petrochenkov** [comment]
>Actually, even this is not true:

That example kinda makes me rethink the prohibition of bare `\r`s.
Raw strings (including `include_str`) cannot use escaping and have no choice than to use bare `\r` if the carriage return needs to be in the string.

Actually, we may want to prohibit only `\r\r\n` rather than all bare `\r`s at loading time, this will already ensure that the file can survive repeated EOL conversions (e.g. by Git).

**Comment by matklad** [comment]
>this will already ensure that the file can survive repeated EOL conversions (e.g. by Git).

It looks like Git treats files with lone CR as binary and doesn't convert them:
https://github.com/git/git/blob/e35b8cb8e212e3557efc565157ceb5cbaaf0d87f/convert.c#L90-L103

I also begin to feel that forbidding lone CR causes more problems that it solves. Sole `\r` does not seem that different from any other weird unicode/ascii junk like `\v`. I'll implement simple `\r\r\n -> \r\n`  conversion, and we can have a separate "weird whitespace" lint if we want to, but it seems to be unlikely that folks will hit this problem in practice.

**Comment by matklad** [comment]
Marking as waiting for review. I've changed this to normalize the string, and I think we should move forward with this regardless of #62948.

@rustbot modify labels: -S-waiting-on-author, +S-waiting-on-review

**Comment by petrochenkov** [comment]
Let's just keep the status quo behavior of `include_str` (at least for now) and not introduce a new third flavor of normalization.

I wanted to write out some possible use cases for various flavor of normalization, but I just need to go and sleep now.

TLDR is that BOM+EOL normalization is the common case requirement for text (`mod m;`, `include!`, `include_str!` (simply long strings, generated strings)) and should attempt to to change `include_str` in this direction and see what breaks (crater, some other crates.io search, live experiment on nightly), but `include_bytes_as_str` reading bytes without any normalization like `include_bytes` and packaging them as `str` instead of `[u8; N]` would also make sense, even if rarely (some ASCII-based binary protocols requiring `\r\n` (do they exist?)? Old WinAPI text controls unable to work with `\n`?).

**Comment by matklad** [comment]
based on https://github.com/rust-lang/rust/pull/63525#issuecomment-521035211

@bors r=petrochenkov

I've removed the normalization commit, but I've added the test that normalization doesn't happen. 

I think the way forward here is to complete #62948 first, and then have a separate discussion/crater runs about `include_str` normalization.

>but include_bytes_as_str reading bytes without any normalization like include_bytes and packaging them as str instead of [u8; N] would also make sense, even if rarely (some ASCII-based binary protocols requiring \r\n (do they exist?)?

I feel this can be build on top of `include_bytes!` with 80% of usability, which seems ok for a rare use-case.

```rust
fn my_text_file_with_crlf() -> &'static str {
    unsafe { 
        std::str::from_utf8_unchecked(&include_bytes!("my_text_file_with_crlf.txt")[..])
    }
}
```
